### PR TITLE
bugfix: optimize redis limit and fix redis page bug

### LINKED ISSUE #4463: redis mode console query globalSession error
- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅱ. Describe what happened
1:redis 控制台分页查询globalSession报错,
2:分页数据偶尔显示错误，多于限制条数
  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```

### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅵ. Environment:

- JDK version :
- Seata version:  
- OS :
- Others:

### PR DESCRIPTION
**Pull request by doubleDimple**
- [ ] I have registered the PR [changes](https://github.com/seata/seata/tree/develop/changes).

### Ⅰ. Describe what this PR did
1:优化redis配置默认查询条数后,均衡查询每个key显示的数据,
2:解决redis分页查询,不再使用模糊key的方式匹配,修改为通过statusKey的方式查询
3:解决控制台redis分页查询,偶尔出现分页数据不准确
### Ⅱ. Does this pull request fix one issue?

fixes #4463 

### Ⅲ. Why don't you add test cases (unit test/integration test)? 
yes ,see RedisTransactionStoreManagerTest

### CODE REVIEW

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 1]
```diff
@@ -419,35 +412,56 @@ public GlobalSession readSession(String xid) {
...
-            if (CollectionUtils.isNotEmpty(list)) {
-                List<String> xids = list.stream().flatMap(ll -> ll.stream()).collect(Collectors.toList());
-                xids.parallelStream().forEach(xid -> {
-                    GlobalSession globalSession = this.readSession(xid, withBranchSessions);
-                    if (globalSession != null) {
-                        globalSessions.add(globalSession);
-                    }
-                });
+            for (String key:statusKeys) {
+                final Long count = jedis.llen(key);
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:451]: long 即可

Comment (a364176773) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:451]: 用pipeline批量llen,再取回,一次网络io即可

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 2]
```diff
@@ -419,35 +412,56 @@ public GlobalSession readSession(String xid) {
...
+
+    }
+
+    /**
+     * query and sort existing values
+     * @param statusKeys
+     * @return
+     */
+    private Map<String,Long> calculateStatuskeysHasData(List<String> statusKeys) {
+        Map<String,Long> keysMap = new ConcurrentHashMap<>();
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:448]: hashmap即可,下面没有多线程,也没有用并行流

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 1 above)
Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:451]: get

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 2 above)
Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:448]: get

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 3]
```diff
@@ -419,35 +411,37 @@ public GlobalSession readSession(String xid) {
...
+        if (targetMap.size() == 0 || logQueryLimit <= 0) {
             return globalSessions;
         }
+        // totalCount
+        final Long totalCount = countByClobalSesisons(statuses);
+
+        List<List<String>> list = new ArrayList<>();
+        dogetXidsForTargetMap(targetMap, 0, limit - 1, logQueryLimit, totalCount, list);
+        if (CollectionUtils.isNotEmpty(list)) {
+            List<String> xids = list.stream().flatMap(ll -> ll.stream()).collect(Collectors.toList());
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:429]: List<String> xids = list.stream().flatMap(Collection::stream).collect(Collectors.toList());

Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:429]: get

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 4]
```diff
@@ -669,30 +691,60 @@ public Long countByClobalSesisons(GlobalStatus[] values) {
...
-        String xid = null;
+    private List<String> convertStatusKeys(GlobalStatus[] statuses) {
+        List<String> statusKeys = new ArrayList<>();
+        for (int i = 0; i < statuses.length; i++) {
+            statusKeys.add(buildGlobalStatus(statuses[i].getCode()));
+        }
+        return statusKeys;
+    }
+
+    private void dogetXidsForTargetMap(Map<String, Long> targetMap, int start, int end, int logQueryLimit,
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:702]: logQueryLimit为什么需要传捡来?外部直接Math.min(totalCount,totalCount),取一个最小进来就行了

**Review by a364176773 — APPROVED** [review]
LGTM

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 5]
```diff
@@ -419,35 +419,42 @@ public GlobalSession readSession(String xid) {
...
-            int size = statusKeys.size();
-            resetLimitQuery = (logQueryLimit / size) == 0 ? logQueryLimit : (logQueryLimit / size);
+    /**
+     * get everyone keys limit
+     * 
+     * @param targetMap
+     * @return
+     */
+    private long resetLogQueryLimit(Map<String, Long> targetMap) {
+        long resetLimitQuery = logQueryLimit;
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:454]: int

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 6]
```diff
@@ -419,35 +419,42 @@ public GlobalSession readSession(String xid) {
...
+        List<GlobalSession> globalSessions = Collections.synchronizedList(new ArrayList<>());
+        List<String> statusKeys = convertStatusKeys(statuses);
+
+        Map<String, Long> targetMap = calculateStatuskeysHasData(statusKeys);
+        long limit = resetLogQueryLimit(targetMap);
+        if (targetMap.size() == 0 || logQueryLimit <= 0) {
             return globalSessions;
         }
+        // totalCount
+        final long totalCount = Math.min(logQueryLimit, countByClobalSesisons(statuses));
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:432]: why use `countByClobalSesisons` but not targetMap's values sum?

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 7]
```diff
@@ -669,30 +690,64 @@ public Long countByClobalSesisons(GlobalStatus[] values) {
...
-                    if (withBranchSessions) {
-                        branchTransactionDOs = this.readBranchSessionByXid(jedis,xid);
-                    }
-                    globalSessions.add(getGlobalSession(globalTransactionDO,branchTransactionDOs, withBranchSessions));
+            for (String key : targetMap.keySet()) {
+                final long sum = listList.stream().mapToLong(value -> value.size()).sum();
+                final long diffCount = totalCount - sum;
+                List<String> list;
+                if (end - start >= diffCount) {
+                    long endNew = end - diffCount;
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:720]: endNew = start + diffCount -1

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 8]
```diff
@@ -669,30 +690,64 @@ public Long countByClobalSesisons(GlobalStatus[] values) {
...
-                    }
-                    List<BranchTransactionDO> branchTransactionDOs = new ArrayList<>();
-                    if (withBranchSessions) {
-                        branchTransactionDOs = this.readBranchSessionByXid(jedis,xid);
-                    }
-                    globalSessions.add(getGlobalSession(globalTransactionDO,branchTransactionDOs, withBranchSessions));
+            for (String key : targetMap.keySet()) {
+                final long sum = listList.stream().mapToLong(value -> value.size()).sum();
+                final long diffCount = totalCount - sum;
+                List<String> list;
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:718]: diffCount <=0; return ;

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 9]
```diff
@@ -419,35 +419,42 @@ public GlobalSession readSession(String xid) {
...
-                        globalSessions.add(globalSession);
-                    }
-                });
-            }
+
+        List<GlobalSession> globalSessions = Collections.synchronizedList(new ArrayList<>());
+        List<String> statusKeys = convertStatusKeys(statuses);
+
+        Map<String, Long> targetMap = calculateStatuskeysHasData(statusKeys);
+        long limit = resetLogQueryLimit(targetMap);
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:427]: int perStatusLimit

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 6 above)
Comment (slievrly) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:432]: totalCount -> queryCount

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 10]
```diff
@@ -669,30 +690,64 @@ public Long countByClobalSesisons(GlobalStatus[] values) {
...
+        long totalCount, List<List<String>> listList) {
+
+        long total = listList.stream().mapToLong(value -> value.size()).sum();
+
+        if (total >= totalCount) {
+            return;
+        }
+        // when start greater than offset(totalCount)
+        if (start >= totalCount) {
+            return;
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:711]: why need recursive? if first access this method totalCount=20  and targetMap.size=4,but during the query operation, part of the transactions ends, and only two transactions matched. 

1.  dogetXidsForTargetMapRecursive(map,0,4,20,listList)  => listList add two element.
2. dogetXidsForTargetMapRecursive(map,5,9,listList) => listList add zero element.
3. dogetXidsForTargetMapRecursive(map,10,14,listList) => listList add zero element.
4. dogetXidsForTargetMapRecursive(map,15,19,listList) => listList add zero element.
5. dogetXidsForTargetMapRecursive(map,20,24,listList) => start >= totalCount return;

Queries for steps 2, 3 and 4 are wasteful and return conditions need to be optimized.

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 11]
```diff
@@ -669,30 +690,64 @@ public Long countByClobalSesisons(GlobalStatus[] values) {
...
+        int pageSize) {
+        List<List<String>> listList = new ArrayList<>();
+        try (Jedis jedis = JedisPooledFactory.getJedisInstance()) {
+            for (String key : targetMap.keySet()) {
+                final List<String> list = jedis.lrange(key, start, end);
+                final long sum = listList.stream().mapToLong(value -> value.size()).sum();
+                if (list.size() > 0 && sum < pageSize) {
+                    listList.add(list);
+                } else {
+                    start = 0;
```
Comment (slievrly) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:745]: why sum > pageSize,  don't add element to listList.

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 6 above)
Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:432]: get

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 5 above)
Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:454]: get

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 7 above)
Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:720]: get

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 8 above)
Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:718]: get

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 9 above)
Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:427]: get

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 11 above)
Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:745]: 这里的pageSize表示totalCount

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 10 above)
Comment (doubleDimple) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:721]: 这里递归主要的作用是补偿每次查询后,数据量不足,需要重新从第一个key,重新指定start和end后,继续查询,对mao执行排序,让数据量最大的排序到前面

**Review by lightClouds917 — APPROVED** [review]
LGTM

**Comment by lightClouds917** [comment]
countByClobalSesisons ->countByGlobalSesisons

**Pull request merged**